### PR TITLE
Support compiling rustc without LLVM (try 2)

### PR DESCRIPTION
**Pull request by bjorn3**
Now doesn't change rustc_driver.

Supersedes #42752

### CODE REVIEW

**Comment by eddyb** [comment]
cc @rust-lang/compiler 
I am personally not in favor of the approach taken here and would rather make `rustc_trans` LLVM-only and extract all the other functionality out of it. There are just too many `#[cfg(llvm)]`.

File: src/bootstrap/config.toml.example [hunk 1]
```diff
@@ -14,6 +14,9 @@
 # =============================================================================
 [llvm]
 
+# Indicates whether rustc will support compilation with LLVM (currently broken)
```
Comment (Keruspe) [inline src/bootstrap/config.toml.example]: The comment seems to imply that the build *with* llvm is broken, not the build without it. Should be the other way round, right?

Comment (bjorn3) [inline src/bootstrap/config.toml.example]: Yes

**Comment by oli-obk** [comment]
> I am personally not in favor of the approach taken here and would rather make rustc_trans LLVM-only and extract all the other functionality out of it. There are just too many #[cfg(llvm)].

So llvm would just be a cargo feature of `rustc_driver` either including `rustc_trans_llvm` or not depending on the feature?

**Comment by eddyb** [comment]
@oli-obk More or less, yes.

**Comment by carols10cents** [comment]
friendly ping to keep this on your radar, @bjorn3!

**Comment by nikomatsakis** [comment]
Without looking in detail, I definitely agree with @eddyb that we should strive to factor things better and avoid quite so many `#[cfg]` gates. @bjorn3, do you need more detailed guidance on how to achieve that?

**Comment by bjorn3** [comment]
@eddyb is this a bit like you mean. If so I will work it out further in a few days.

**Comment by alexcrichton** [comment]
Looks like this may be ready for another round of review @eddyb!

@bjorn3 it also looks like there may be rebase conflicts here now.

File: src/librustc_driver/Cargo.toml [hunk 2]
```diff
@@ -29,9 +29,14 @@ rustc_plugin = { path = "../librustc_plugin" }
...
 rustc_typeck = { path = "../librustc_typeck" }
 serialize = { path = "../libserialize" }
 syntax = { path = "../libsyntax" }
 syntax_ext = { path = "../libsyntax_ext" }
 syntax_pos = { path = "../libsyntax_pos" }
+
+ar = "0.3.0"
+
+[features]
+llvm = ["rustc_trans"]
```
Comment (eddyb) [inline src/librustc_driver/Cargo.toml]: Missing newline at the end of file.

**Review by eddyb — APPROVED** [review]
Now this looks much more streamlined!

**Comment by eddyb** [comment]
@bjorn3 Ahh how I wish I had a way to un-dismiss github notifications... I keep losing track of some I've gone through on my phone 😞. @alexcrichton Thanks for the ping!

File: src/librustc_driver/Cargo.toml (hunk 2 above)
Comment (bjorn3) [inline src/librustc_driver/Cargo.toml]: Fixed

**Comment by bjorn3** [comment]
Rebased

**Comment by eddyb** [comment]
@bjorn3 Can you squash the commits to remove the back-and-forth (X followed by revert X)?

File: src/librustc_driver/lib.rs [hunk 3]
```diff
@@ -151,6 +157,44 @@ pub fn run<F>(run_compiler: F) -> isize
...
+pub struct NoLLvmMetadataLoader;
+
+#[cfg(not(feature="llvm"))]
+type LlvmMetadataLoader = NoLLvmMetadataLoader;
+#[cfg(feature="llvm")]
+type LLvmMetadataLoader = rustc_trans::LlvmMetadataLoader;
+
+#[cfg(not(feature="llvm"))]
+impl MetadataLoader for NoLLvmMetadataLoader {
+    fn get_rlib_metadata(&self, _: &Target, filename: &Path) -> Result<ErasedBoxRef<[u8]>, String> {
```
Comment (eddyb) [inline src/librustc_driver/lib.rs]: Was this copied wholesale from `rustc_trans`? Does that mean `ar::Archive` is independent of LLVM?
That could make `rustc_metadata` target-agnostic in the non-dynamic-library case!
cc @alexcrichton

Comment (bjorn3) [inline src/librustc_driver/lib.rs]: `ar` is a rust only `.a` reader and writer.

Comment (eddyb) [inline src/librustc_driver/lib.rs]: Then it should be depended on by `rustc_metadata` and `get_rlib_metadata` shouldn't exist which means `loader: Box<MetadataLoader>` in `rustc_metadata` becomes just:
```rust
get_dylib_metadata: fn(&Target,  &Path) -> Result<ErasedBoxRef<[u8]>, String>
```
When not available, one can just pass in `|_, _| Err(String::from("unsupported"))`.

**Comment by bjorn3** [comment]
Squashed

**Comment by shepmaster** [comment]
This is marked as a WIP which means we are still waiting on @bjorn3 to finish the PR.

**Comment by eddyb** [comment]
See https://github.com/rust-lang/rust/pull/42932#discussion_r130300977 for what is left here.

**Comment by eddyb** [comment]
@bjorn3 This looks great! Can you remove `[WIP]` from the title and rebase?

**Comment by bjorn3** [comment]
Done

**Comment by eddyb** [comment]
cc @oli-obk This PR allows building a `rustc` which doesn't contain LLVM but can read `rlib` files.
I suppose we still need to move the output of the `rlib`s out of `rustc_trans` but I believe it is feasible to do so and have a miri-only compiler, which could, for example, run in the browser.

**Comment by bjorn3** [comment]
Wait! Need to make a change so that rustc_driver actually compiles without LLVM.

**Comment by eddyb** [comment]
@bjorn3 You can force push at any time before the PR starts getting built and it will leave the queue.

**Comment by oli-obk** [comment]
I'm so excited about this change!

**Comment by bjorn3** [comment]
Trying it out locally with llvm disabled.

**Comment by eddyb** [comment]
(hehe that @bors bug isn't fixed yet)

**Comment by bjorn3** [comment]
Should be ready now.

**Comment by bjorn3** [comment]
```
[00:17:04]    Compiling std_unicode v0.0.0 (file:///checkout/src/libstd_unicode)
[00:17:04] error[E0463]: can't find crate for `core`
[00:17:04] 
[00:17:04] error: aborting due to previous error
[00:17:04] 
[00:17:04] error: Could not compile `libc`.
```

**Comment by bjorn3** [comment]
Doesn't fail locally (mac os)

**Comment by alexcrichton** [comment]
Hm so looking over this it looks like this is changing how we're reading rlib metadata? The current construction is done specifically for speed as LLVM's ar parsing/reading is much speedier than many other solutions. Are we sure the `ar` crate is sufficiently fast enough? It initially looks that it is not, as one big difference is using memory maps vs actually reading the file into memory, which can make a huge difference here.

Also from a maintainability perspective, this seems like the kind of patch that'd bitrot almost instantly? Is there something we can do to avoid the `#[cfg]` jungle that will be very difficult to maintain over time?

**Comment by eddyb** [comment]
@alexcrichton OH! Uhm, when I asked about it it sounded like we already moved away from using the LLVM support - but if that's not true, then the "Move rlib metadata reading code to rustc_metadata and replace MetadataLoader with a function pointer" commit was a mistake...

**Comment by bjorn3** [comment]
Travis passes.

**Comment by eddyb** [comment]
@bjorn3 Please squash the original commits with the reverts.

**Comment by bjorn3** [comment]
Squashed and travis passes

**Comment by alexcrichton** [comment]
@eddyb and @bjorn3 did y'all have no comments about the maintainability aspect above? Specifically:

> Also from a maintainability perspective, this seems like the kind of patch that'd bitrot almost instantly? Is there something we can do to avoid the #[cfg] jungle that will be very difficult to maintain over time?

**Comment by eddyb** [comment]
@alexcrichton So I took a look at the diff to count them, remembering only 5-6... and now there's a lot more. And they're mostly in https://github.com/rust-lang/rust/pull/42932/commits/b7314c7cafc657afa6b98eb9c4958e56a82129bb "Actually make rustc_driver compile without llvm".

**Comment by eddyb** [comment]
So that is definitely my fault, and if @alexcrichton wants this reverted then I'm on board.
@bjorn3 What are you willing to do here? I know I've asked for a lot during this PR.

**Comment by bjorn3** [comment]
Im writing a pr to remove most of the cfg's. It will likely be ready tomorrow.

**Comment by alexcrichton** [comment]
Thanks @bjorn3! Feel free to cc or r? me on that

**Pull request merged**